pkg/notes/options/options.go

```diff
@@ -0,0 +1,376 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package options
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/google/go-github/v28/github"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/oauth2"
+
+	"k8s.io/release/pkg/git"
+	"k8s.io/release/pkg/notes/client"
+	"k8s.io/release/pkg/notes/client/clientfakes"
```

Comment:
I strongly opt against mixing test- and prod-code.

If I understand you proposed change correctly, all you'd need to do to make that happen is to implement two or more new clients. E.g. a `notes.RecordingClient` and a `notes.ReplayingClient`. Both would need to implement the [`Client` interface][clientInt]. When you set up the [`Gatherer`][gatherer], you can decide, based on your command line flag/options/... which client you hand in:
- a [wrapped github client][wrappedClient] (the "default" client)
- a `RecordingClient`
- a `ReplayingClient`

The `RecordingClient` could just wrap a real client `client := notes.RecordingClient{Client: client, Recordings: "/some/dir"}` and intercept & save the responses, the `ReplayingClient` just needs the dump dir `client := notes.ReplayingClient{Recordings: "/some/dir"}`. Those different clients could be nicely tested in isolation and we'd have a nice separation of concerns.

The options package and struct do way too much for my taste anyway, I don't think we should cram more functionality onto it.

Also, in general this seems like a test concern and I am not sure if we'd want to expose that via the `krel` (or any other) binary, but we should maybe see to make those test cases / create a test harness.  But if you think exposing that via the binaries is useful, I certainly won't block that!


[clientInt]: https://github.com/kubernetes/release/blob/2c9ce919c182eb72801775c036861a29aa1e88e9/pkg/notes/client.go#L31-L35
[gatherer]: https://github.com/kubernetes/release/blob/2c9ce919c182eb72801775c036861a29aa1e88e9/cmd/krel/cmd/changelog.go#L215-L220
[wrappedClient]: https://github.com/kubernetes/release/blob/2c9ce919c182eb72801775c036861a29aa1e88e9/pkg/notes/client.go#L41-L43

Reply:
> I strongly opt against mixing test- and prod-code.
> 
> If I understand you proposed change correctly, all you'd need to do to make that happen is to implement two new clients. E.g. a `notes.RecordingClient` and a `notes.ReplayingClient`. Both would need to implement the [`Client` interface](https://github.com/kubernetes/release/blob/2c9ce919c182eb72801775c036861a29aa1e88e9/pkg/notes/client.go#L31-L35). When you set up the [`Gatherer`](https://github.com/kubernetes/release/blob/2c9ce919c182eb72801775c036861a29aa1e88e9/cmd/krel/cmd/changelog.go#L215-L220), you can decide, based on your command line flag/options/... which client you hand in:
> 
> * a [wrapped github client](https://github.com/kubernetes/release/blob/2c9ce919c182eb72801775c036861a29aa1e88e9/pkg/notes/client.go#L41-L43) (the "default" client)
> * a `RecordingClient`
> * a `ReplayingClient`
>
> The `RecordingClient` could just wrap a real client `client := notes.RecordingClient{Client: client, Recordings: "/some/dir"}` and intercept & save the responses, the `ReplayingClient` just needs the dump dir `client := notes.ReplayingClient{Recordings: "/some/dir"}`. Those different clients could be nicely tested in isolation and we'd have a nice separation of concerns.
> 
> The options package and struct do way too much for my taste anyway, I don't think we should cram more functionality onto it.

Alright, thank you for the input. I separated both implementations and putt them into a dedicated `recordreplay` package. 

> Also, in general this seems like a test concern and I am not sure if we'd want to expose that via the `krel` (or any other) binary, but we should maybe see to make those test cases / create a test harness. But if you think exposing that via the binaries is useful, I certainly won't block that!

Yeah I thought it would be not only good for testing, but also for development: If I want to change the layout/parsing of the release notes it would be good to be able to fastly replay the latest recording without having the need for any GitHub API access. WDYT?